feat: add C++ modules support

JohelEGP opened the pull request:
Resolves #7.

JohelEGP commented on src/mp_units.cpp:
```diff
@@ -0,0 +1,74 @@
+
+export module mp_units;
+
+export import <array>;
+export import <compare>;
+import <optional>;
+import <ratio>;
+import <tuple>;
```
Other than `<compare>`, I'm not sure why these are necessary. They come from the currently standalone `magnitude.h`.

mpusz replied:
I am not sure what you mean here?

JohelEGP replied:
Without these, compiling the module during the processed `magnitude.h` or `prime.h` breaks, and for the exported `<array>`, also on use of the static `std::array` data member in the tests. No other `mp_units` header required me to `import` a C++ standard library header (besides the exported `<compare>` on uses of `<=>` by importers).

JohelEGP commented on test/unit_test/static/quantity_point_kind_test.cpp:
```diff
@@ -460,7 +469,7 @@ static_assert(!std::is_assignable_v<abscissa<metre, int>, abscissa<millimetre, i
 #if !defined(UNITS_COMP_MSVC) || defined(NDEBUG)
 static_assert([]() {
-  const width<metre, int> w(1 * m);
+  [[maybe_unused]] const width<metre, int> w(1 * m);
```
```suggestion
  const width<metre, int> w(1 * m);
```
This slipped.

JohelEGP commented on test/unit_test/static/quantity_point_test.cpp:
```diff
@@ -37,7 +42,7 @@ namespace {
 
 using namespace units;
 using namespace isq::si;
-using namespace references;
+using namespace isq::si::references;
```
Why did this work before? Oh, the `units::reference` in `generic/angle.h` wasn't visible before.

JohelEGP reviewed (commented):
I missed exporting the aliases in `units::aliases::isq::si`.

JohelEGP commented:
Limitations have been documented: https://github.com/JohelEGP/jegp.cmake_modules/#limitations.

mpusz commented on src/core/include/units/bits/external/fixed_string.h:
```diff
@@ -23,7 +23,6 @@
 #pragma once
 
 #include <units/bits/external/hacks.h>  // IWYU pragma: keep
-// TODO use <algorithm> when moved to C++20 modules (parsing takes too long for each translation unit)
```
`#if UNITS_MODULES`?

JohelEGP replied:
I guess it works by coincidence. Before some refactoring, an `import <algorithm>;` used to be visible.

mpusz replied:
I meant that the `algorithm` header should be only included/imported then `UNITS_MODULES` is `true`, right?

JohelEGP replied:
That's right.

JohelEGP replied:
I left this alone for now.

mpusz commented on src/mp_units.cpp:
```diff
@@ -0,0 +1,74 @@
+#include <units/reference.h>
+#include <units/symbol_text.h>
+#include <units/unit.h>
+
```
What about `core-fmt` and `core-io`?

JohelEGP replied:
I should add a `// TODO` or do it.

mpusz commented on src/mp_units.cpp:
```diff
@@ -0,0 +1,74 @@
+#include <units/unit.h>
+
+// systems
+#include <units/isq/dimensions.h>
+#include <units/isq/iec80000/iec80000.h>
+#include <units/isq/natural/natural.h>
+#include <units/isq/si/cgs/cgs.h>
+#include <units/isq/si/hep/hep.h>
+#include <units/isq/si/iau/iau.h>
+#include <units/isq/si/si.h>
+#include <units/isq/si/typographic/typographic.h>
+#include <units/isq/si/uscs/uscs.h>
```
Is it possible to support submodules already?

JohelEGP replied:
Yes. However, I decided against that. Up to `mp_units.core` and `mp_units.isq.dimensions` and perhaps some of the lighter systems, everything compiled under a second. But `mp_units.isq.si` and some others took more, bringing the total to almost 3 minutes.

The current single module `mp_units` takes almost 1 minute to compile. All times can be halved if I could find a way to compile the object and CMI in parallel while still requiring the CMI to recompile if the object needs to recompile, which is currently done as a direct dependency and thus are not parallelized.

mpusz replied:
Why do separate modules compile 3 times slower than the one big module? Some things like IO will probably be defined as separate modules as it might be a no-go for some customers to use that.

mpusz replied:
Some things like IO will probably have to be defined as separate modules as it might be a no-go for some customers to use that. Also, I would like to have at least core being separate from the systems as well (in case someone wants to just use the framework and not systems predefined by us).

JohelEGP replied:
I don't know and can only speculate. Perhaps each has to do some amount of repeated instantiations that just doesn't happen in One Big Module because they're memoized.

JohelEGP replied:
The double compilation is not required after the realization stated at https://github.com/mpusz/units/pull/350#discussion_r1126285956. I'll look into https://github.com/mpusz/units/pull/350#discussion_r841088509.

JohelEGP replied:
I did some measurements.
Compiled item | Time
--- | ---
Module `mp_units` | 33 s
Static tests (`#include`) | 1 m 11 s
Static tests (`mp_units`) | 3 m 34 s

I compiled a target with the sources of the static tests in both modes (`#include`, `mp_units`)
with property `CXX_COMPILER_LAUNCHER` `"${CMAKE_COMMAND};-E;time"`.
For good measure, I used `-j1`.

Compiled item | `#include` time | Module `mp_units` time
--- | --- | ---
[  3%] fixed_string_test.cpp.o | Elapsed time: 1 s. (time), 0.000485 s. (clock) | Elapsed time: 4 s. (time), 0.00048 s. (clock)
[  6%] symbol_text_test.cpp.o | Elapsed time: 1 s. (time), 0.000436 s. (clock) | Elapsed time: 4 s. (time), 0.000775 s. (clock)
[ 10%] ratio_test.cpp.o | Elapsed time: 1 s. (time), 0.00061 s. (clock) | Elapsed time: 9 s. (time), 0.00046 s. (clock)
[ 13%] prime_test.cpp.o | Elapsed time: 1 s. (time), 0.000454 s. (clock) | Elapsed time: 1 s. (time), 0.000537 s. (clock)
[ 16%] type_list_test.cpp.o | Elapsed time: 2 s. (time), 0.000521 s. (clock) | Elapsed time: 11 s. (time), 0.00057 s. (clock)
[ 20%] dimension_op_test.cpp.o | Elapsed time: 2 s. (time), 0.000505 s. (clock) | Elapsed time: 12 s. (time), 0.000696 s. (clock)
[ 23%] chrono_test.cpp.o | Elapsed time: 4 s. (time), 0.000527 s. (clock) | Elapsed time: 16 s. (time), 0.00055 s. (clock)
[ 26%] custom_rep_test_min_expl.cpp.o | Elapsed time: 2 s. (time), 0.000508 s. (clock) | Elapsed time: 14 s. (time), 0.000586 s. (clock)
[ 30%] custom_rep_test_min_impl.cpp.o | Elapsed time: 3 s. (time), 0.000521 s. (clock) | Elapsed time: 16 s. (time), 0.000421 s. (clock)
[ 33%] custom_unit_test.cpp.o | Elapsed time: 6 s. (time), 0.000467 s. (clock) | Elapsed time: 15 s. (time), 0.000497 s. (clock)
[ 36%] dimensions_concepts_test.cpp.o | Elapsed time: 19 s. (time), 0.000459 s. (clock) | Elapsed time: 12 s. (time), 0.000589 s. (clock)
[ 40%] kind_test.cpp.o | Elapsed time: 3 s. (time), 0.00046 s. (clock) | Elapsed time: 1 s. (time), 0.000487 s. (clock)
[ 43%] references_test.cpp.o | Elapsed time: 3 s. (time), 0.000495 s. (clock) | Elapsed time: 14 s. (time), 0.000592 s. (clock)
[ 46%] si_test.cpp.o | Elapsed time: 22 s. (time), 0.000531 s. (clock) | Elapsed time: 20 s. (time), 0.000504 s. (clock)
[ 50%] unit_test.cpp.o | Elapsed time: 2 s. (time), 0.000612 s. (clock) | Elapsed time: 11 s. (time), 0.000525 s. (clock)
[ 53%] iec80000_test.cpp.o | Elapsed time: 3 s. (time), 0.000444 s. (clock) | Elapsed time: 14 s. (time), 0.000478 s. (clock)
[ 56%] cgs_test.cpp.o | Elapsed time: 5 s. (time), 0.000506 s. (clock) | Elapsed time: 16 s. (time), 0.000554 s. (clock)
[ 60%] si_cgs_test.cpp.o | Elapsed time: 7 s. (time), 0.000552 s. (clock) | Elapsed time: 18 s. (time), 0.000453 s. (clock)
[ 63%] fps_test.cpp.o | Elapsed time: 8 s. (time), 0.000485 s. (clock) | Elapsed time: 17 s. (time), 0.000415 s. (clock)
[ 66%] si_fps_test.cpp.o | Elapsed time: 13 s. (time), 0.000521 s. (clock) | Elapsed time: 18 s. (time), 0.000474 s. (clock)
[ 70%] si_hep_test.cpp.o | Elapsed time: 13 s. (time), 0.000592 s. (clock) | Elapsed time: 14 s. (time), 0.000544 s. (clock)
[ 73%] us_test.cpp.o | Elapsed time: 3 s. (time), 0.000477 s. (clock) | Elapsed time: 15 s. (time), 0.000514 s. (clock)
[ 76%] concepts_test.cpp.o | Elapsed time: 5 s. (time), 0.000466 s. (clock) | Elapsed time: 18 s. (time), 0.000449 s. (clock)
[ 80%] math_test.cpp.o | Elapsed time: 5 s. (time), 0.000456 s. (clock) | Elapsed time: 13 s. (time), 0.000471 s. (clock)
[ 83%] point_origin_test.cpp.o | Elapsed time: 5 s. (time), 0.000476 s. (clock) | Elapsed time: 1 s. (time), 0.000503 s. (clock)
[ 86%] angle_test.cpp.o | Elapsed time: 2 s. (time), 0.000873 s. (clock) | Elapsed time: 14 s. (time), 0.00044 s. (clock)
[ 90%] quantity_kind_test.cpp.o | Elapsed time: 9 s. (time), 0.000499 s. (clock) | Elapsed time: 23 s. (time), 0.000902 s. (clock)
[ 93%] quantity_point_kind_test.cpp.o | Elapsed time: 8 s. (time), 0.000565 s. (clock) | Elapsed time: 20 s. (time), 0.00084 s. (clock)
[ 96%] quantity_point_test.cpp.o | Elapsed time: 8 s. (time), 0.000472 s. (clock) | Elapsed time: 15 s. (time), 0.000584 s. (clock)
[100%] quantity_test.cpp.o | Elapsed time: 8 s. (time), 0.000499 s. (clock) | Elapsed time: 19 s. (time), 0.000785 s. (clock)

mpusz replied:
WOW, that is not what I expected. I thought that the user's code would benefit from having all the instantiations done already in the module. But it seems that modules actually provide much worse performance to the users :-(

JohelEGP replied:
`mp_units.core` is a win. But as mentioned at https://github.com/mpusz/units/pull/350#discussion_r840593083, there's little else I can do. Most systems require SI, which is heavy.

Compiled item | `#include` | `mp_units.core`
--- | --- | ---
`angle_test` | 2 s | 1 s
`dimension_op_test` | 2 s | 0 s
`fixed_string_test` | 1 s | 0 s
`prime_test` | 1 s | 1 s
`ratio_test` | 1 s | 0 s
`symbol_text_test` | 1 s | 0 s
`type_list_test` | 2 s | 0 s
`unit_test` | 2 s | 1 s

mpusz replied:
But why preprocessing SI from scratch all over again in a header file is faster than reusing already parsed info in a module?

JohelEGP replied:
My guess is that accessing what's been written to the PCM (Clang's CMI, Compiled Module Interface) is slower to access than what the compiler parses from textual inclusion, when the target PCM is a heavy template library, as in this case.

JohelEGP replied:
Opened https://github.com/llvm/llvm-project/issues/61226.

JohelEGP replied:
The LLVM issue seems fixed.

mpusz commented on test/unit_test/static/chrono_test.cpp:
```diff
@@ -67,12 +73,12 @@ static_assert(!std::constructible_from<time_point<std::chrono::system_clock, si:
-static_assert(!std::convertible_to<std::chrono::hours, si::time<si::second>>);
-static_assert(std::constructible_from<si::time<si::hour>, std::chrono::seconds>);
-static_assert(!std::convertible_to<std::chrono::seconds, si::time<si::hour>>);
+static_assert(std::constructible_from<si::time<si::second, double>, std::chrono::seconds>);
+static_assert(!std::convertible_to<std::chrono::seconds, si::time<si::second, double>>);
+static_assert(std::constructible_from<si::time<si::second, double>, std::chrono::hours>);
+static_assert(!std::convertible_to<std::chrono::hours, si::time<si::second, double>>);
+static_assert(std::constructible_from<si::time<si::hour, double>, std::chrono::seconds>);
+static_assert(!std::convertible_to<std::chrono::seconds, si::time<si::hour, double>>);
```
Defaults for template arguments do not work with modules?

JohelEGP replied:
That's one of the issues listed in #7: https://github.com/llvm/llvm-project/issues/54340.

JohelEGP replied:
This is not required after the realization stated at https://github.com/mpusz/units/pull/350#discussion_r1126285956.

mpusz commented on src/mp_units.cpp:
```diff
@@ -0,0 +1,74 @@
+// The MIT License (MIT)
```
It seems that this is no longer a header-only library, right? It will produce different artifacts for different compilers. So probably we should get rid of the: https://github.com/mpusz/units/blob/70586138d9590076a3b46b12133956cf97dcfc41/conanfile.py#L145-L146 in case modules are being used (additional Conan option needed).

JohelEGP replied:
Just not "only". The header-only library targets are still there, and are even used by `mp-units::module` through its linked library `mp-units::systems`.

mpusz replied:
Yes, but as long as there is at least one artifact that is compiler-dependent, Conan has to generate a different ID for such configurations. This is why we need a module-specific ON/OFF option in Conan and when it is on we should not use `self.info.header_only()` as this forces generation of the same identifier for the package (no matter what the compiler settings are).

JohelEGP replied:
I see. Do I just need to remove those lines?

mpusz replied:
Yes, add a new Conan option and make a branch that will run it only when modules are not being used.

mpusz replied:
Also please add docs here: https://mpusz.github.io/units/usage.html#conan-options for both Conan and CMake.

mpusz commented on test/unit_test/static/quantity_kind_test.cpp:
```diff
@@ -32,7 +34,10 @@
+import mp_units;
+import <chrono>;
+#endif
 #include <functional>
 #include <limits>
 #include <type_traits>
```
Why only `<chrono>` is imported while other STD headers are `#include`d?

JohelEGP replied:
No particular reason. This highlights that `chrono.h` guarantees `<chrono>` is `#include`d, whereas `mp_units` doesn't export `<chrono>`. Do you think it should?

mpusz replied:
Aren't `#include`s also not exported? It would be nice to be consistent. Also is there any difference in preprocessor parsing overhead between `import <header>` and `#include <header>` or is it equivalent?

JohelEGP replied:
> Aren't `#include`s also not exported?

That's right.



> Also is there any difference in preprocessor parsing overhead between `import <header>` and `#include <header>` or is it equivalent?

It depends on the implementation. Clang seems to translate `#include`s to `import`. GCC doesn't, and attempting to include standard headers while importing `mp_units` would break it due to declarations being both reachable from `mp_units` (because in the current state, `#include`s happen outside the GFM, but that's OK because Clang translates them to `import`s) as well as from where they were `#include`d.

As for the parsing overhead, I don't have a direct answer. But on my machine, compiling a module that just exports all importable C++ standard library headers required by `mp_units` takes like 0.25s, whereas compiling the `mp_units.core` that imports such a module takes less than 1s, so it's negligible.

JohelEGP replied:
For now, I `export import std;` in `module mp_units;`.

mpusz reviewed (commented):
Huge work! Thank you!!!

JohelEGP commented on src/core/include/units/bits/basic_concepts.h:
```diff
@@ -38,19 +38,22 @@
 // Prefix
 namespace detail {
 
+UNITS_LLVM_EXPORT
```
Yes, all added macros prefixed with `UNITS_LLVM_` are about LLVM reachability bugs, and should go away when they're fixed and support for versions of Clang that require them is dropped.

JohelEGP replied:
Turns out it was because I was using Clang modules, which is different from Clang's implementation of C++ modules. The former was documented and the latter not, so I thought they were the same. Now you can find the latter's here: https://clang.llvm.org/docs/StandardCPlusPlusModules.html.

JohelEGP commented on src/core/include/units/bits/basic_concepts.h:
```diff
@@ -129,12 +138,14 @@ inline constexpr bool is_exponent = false;
 // DerivedDimension
 namespace detail {
 
+UNITS_EXPORT
```
That's because some tests use them. I did entertain the though to instead use something like `UNITS_TESTS_EXPORT` for clarity.

mpusz replied:
Maybe it should be a different module in the end that will export its definitions to be used in unit tests but will only be imported (without exporting) by the main library module?

JohelEGP replied:
I don't think that's possible due to how interdependent some headers are. I'd entertain that refactor when modules support is more mature. Attempting this now risks delaying modules support for other compilers.

JohelEGP replied:
For now, I export everything.

JohelEGP commented:
Apparently, not all is lost, as it might be possible to make the downcasting facility work with Clang modules: https://godbolt.org/z/jenGb5hM8. I just need to figure out why this works and attempt the same in mpusz/units.

mpusz commented:
@JohelEGP, is this PR ready to merge (besides merge conflicts)? I thought that you still wanted to add to it.

JohelEGP commented:
I have to address some of the comments above.

JohelEGP commented:
While I work on further integration, this is how I have made use of this PR:
```CMake
  ExternalProject_Add(
    mp-units
    GIT_REPOSITORY "${WAARUDO_mp-units_GIT_REPOSITORY}"
    GIT_TAG "${WAARUDO_mp-units_GIT_TAG}"
    CMAKE_ARGS ${mp_units_args} -DUNITS_AS_MODULES=TRUE -DUNITS_USE_LIBFMT=FALSE
    CMAKE_CACHE_ARGS ${cxx_args} ${jegp_args} ${waarudo_args}
    SOURCE_SUBDIR "src"
    CONFIGURE_HANDLED_BY_BUILD TRUE
    BUILD_COMMAND "${CMAKE_COMMAND}" --build <BINARY_DIR> --target mp-units-module
    INSTALL_COMMAND ""
    DEPENDS jegp.cmake_modules gsl-lite Waarudo_std
    EXCLUDE_FROM_ALL TRUE)
```
This adds https://github.com/JohelEGP/jegp.cmake_modules/#jegpaddmodule to `CMAKE_MODULE_PATH`.

JohelEGP commented:
<https://github.com/Kitware/CMake/blob/v3.27.4/Help/dev/experimental.rst#c20-module-apis>
explains how to use CMake's experimental support for C++ modules.
Change the `v3.27.4` in the link for your CMake version.

Using Clang 18, I'm afraid there seems to be a bug:
<details><summary>Error</summary>

```output
/home/johel/Documents/C++/Forks/mpusz/units/test/unit_test/static/quantity_test.cpp:52:22: error: constraints not satisfied for class template 'quantity' [with R = mp_units::si::metre{{}}, Rep = short]
   52 | static_assert(sizeof(quantity<si::metre, short>) == sizeof(short));
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/quantity.h:85:28: note: because 'RepresentationOf<short, get_quantity_spec(mp_units::si::metre{{}}).character>' evaluated to false
   85 | template<Reference auto R, RepresentationOf<get_quantity_spec(R).character> Rep = double>
      |                            ^
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/bits/representation_concepts.h:82:28: note: because 'short' does not satisfy 'Representation'
   82 | concept RepresentationOf = Representation<T> && ((Ch == quantity_character::scalar && is_scalar<T>) ||
      |                            ^
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/bits/representation_concepts.h:79:93: note: because 'short' does not satisfy 'Scalable'
   79 | concept Representation = (is_scalar<T> || is_vector<T> || is_tensor<T>)&&std::regular<T> && detail::Scalable<T>;
      |                                                                                             ^
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/bits/representation_concepts.h:70:3: note: because 'short' does not satisfy 'CastableNumber'
   70 |   CastableNumber<T> ||
      |   ^
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/bits/representation_concepts.h:65:62: note: because 'std::common_type_t<short, std::intmax_t>' (aka 'long') does not satisfy 'ScalableNumber'
   65 | concept CastableNumber = CommonTypeWith<T, std::intmax_t> && ScalableNumber<std::common_type_t<T, std::intmax_t>>;
      |                                                              ^
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/bits/representation_concepts.h:62:3: note: because 'std::regular_invocable<std::multiplies<>, long, long>' evaluated to false
   62 |   std::regular_invocable<std::multiplies<>, T, U> && std::regular_invocable<std::divides<>, T, U>;
      |   ^
/home/johel/root/clang-main/bin/../include/c++/v1/__concepts/invocable.h:34:29: note: because 'invocable<std::multiplies<void>, long, long>' evaluated to false
   34 | concept regular_invocable = invocable<_Fn, _Args...>;
      |                             ^
/home/johel/root/clang-main/bin/../include/c++/v1/__concepts/invocable.h:28:3: note: because 'std::invoke(std::forward<_Fn>(__fn), std::forward<_Args>(__args)...)' would be invalid: no matching function for call to 'invoke'
   28 |   _VSTD::invoke(_VSTD::forward<_Fn>(__fn), _VSTD::forward<_Args>(__args)...); // not required to be equality preserving
      |   ^
/home/johel/root/clang-main/bin/../include/c++/v1/__config:805:17: note: expanded from macro '_VSTD'
  805 | #  define _VSTD std
      |                 ^
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/bits/representation_concepts.h:71:24: note: and 'typename T::value_type' would be invalid: type 'short' cannot be used prior to '::' because it has no members
   71 |   (requires { typename T::value_type; } && CastableNumber<typename T::value_type> &&
      |                        ^
/home/johel/Documents/C++/Forks/mpusz/units/src/core/include/mp-units/bits/representation_concepts.h:73:24: note: and 'typename T::element_type' would be invalid: type 'short' cannot be used prior to '::' because it has no members
   73 |   (requires { typename T::element_type; } && CastableNumber<std::remove_reference_t<typename T::element_type>> &&
      |                        ^
```

</details>

JohelEGP commented:
This is how I tested this (using [CPM][] instead of Conan).
In the CMake buildsystem:
```CMake
CPMAddPackage("gh:gsl-lite/gsl-lite#4b5e9ab7474841fc2d7efc2e0064ef81785535d1")
CPMAddPackage("gh:fmtlib/fmt#e8259c5298513e8cdbff05ce01c46c684fe758d8")
CPMAddPackage("gh:catchorg/Catch2#3a5cde55b7a27a6d94ce994a8362b2425211bd5e")
execute_process(COMMAND "${CMAKE_COMMAND}" -S "${CPM_Catch2_SOURCE}" -B Catch2
                WORKING_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}" COMMAND_ERROR_IS_FATAL ANY)
set(Catch2_DIR "${CMAKE_CURRENT_BINARY_DIR}/Catch2")
list(APPEND CMAKE_MODULE_PATH "${CPM_Catch2_SOURCE}/extras")
CPMAddPackage(
  NAME "mp-units"
  VERSION "modules"
  GITHUB_REPOSITORY "JohelEGP/mp-units"
  OPTIONS "MP_UNITS_USE_LIBFMT TRUE"
  SOURCE_SUBDIR "src")
add_subdirectory("${CPM_mp-units_SOURCE}/test" "mp-units-test")
add_subdirectory("${CPM_mp-units_SOURCE}/example" "mp-units-example")
```

[CPM]: https://github.com/cpm-cmake/CPM.cmake

mpusz commented on example/glide_computer_lib/CMakeLists.txt:
```diff
@@ -25,3 +25,7 @@ cmake_minimum_required(VERSION 3.2)
 add_library(glide_computer_lib STATIC glide_computer_lib.cpp include/glide_computer_lib.h)
 target_link_libraries(glide_computer_lib PRIVATE mp-units::core-fmt PUBLIC mp-units::si mp-units::utility example_utils)
```
Should we still link those `mp-unit::XXX` targets when a module target is linked?

JohelEGP replied:
Yes, because `mp-units::modules` is only linked when it exists.

JohelEGP replied:
That does sounds more correct.

JohelEGP replied:
Unfortunately, it's not straight forward.
The tests and examples use macros from the implementation of different headers.

mpusz commented on src/CMakeLists.txt:
```diff
@@ -51,6 +51,20 @@ target_link_libraries(
+    )
+    target_link_libraries(mp-units-modules PRIVATE mp-units::mp-units)
+    target_include_directories(
+        mp-units-modules ${unitsAsSystem} INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/core/include>
```
Why only `core` includes are added and not the rest of them? When we link with `mp-units` target that contains all the library shouldn't we also expose all the headers as well?

JohelEGP replied:
This one is for `mp_units/quantity_spec_macro.h`.
The others, `mp_units/bits/external/hacks.h` and `mp_units/bits/fmt_hacks.h`, are not part of the API.

JohelEGP replied:
> When we link with `mp-units` target that contains all the library shouldn't we also expose all the headers as well?

A consumer of the `mp_units` module shouldn't care about other headers besides `mp_units/quantity_spec_macro.h`.

mpusz replied:
Maybe this file should be moved to a special subdirectory like `module_include` and only this path should be exposed in a module?

mpusz replied:
What about the installation part? Did you disable copying of the remaining header files for the module build?

JohelEGP replied:
I agree with moving it standalone.
That way, a consumer of the module `mp_units` wouldn't be able to include other `core` headers.
But `core` would need this header, as I moved it out of `quantity_spec.h`.
However, then we'd also have to update compiler-explorer.

JohelEGP replied:
> What about the installation part? Did you disable copying of the remaining header files for the module build?

To do that, and still support consumers of the header project, there'd be some complications.
I think it's just better to continue providing the header library regardless of support for C++ modules.

mpusz replied:
I will update Compiler Explorer after merge. No worries.

mpusz replied:
> To do that, and still support consumers of the header project, there'd be some complications.

What is the issue here? Can't we do it with branching in CMake?

JohelEGP replied:
I won't know until I try it.
But I imagine we'd have to make something like
`find_package(mp_units COMPONENTS modules)` not provide the header-only library.
That may be doable.

mpusz replied:
I am not so sure about it. I would like to get to a point that if I do:
- `cmake install -DHEADERS ...` I will see all the headers copied to the target directory
- `cmake install -DMODULE ...` I will have the BMI and only the `quantity_spec_macro.h` copied to the target directory

Of course, the name of the above defines is just for this example.

The same should be the case for Conan packages for both options (because there should be an option like `module: true/false` exposed from Conan).

JohelEGP replied:
We'll see how that works.
Installing modules is not yet supported.
And you need to unreleased CMake v3.28.0 to consume imported modules.

I do remember that before, the plan was to install the sources, and have the consumers build the BMI.
But I think the recent talks suggest that the direction will be to install a BMI, as you suggest.

mpusz commented on src/CMakeLists.txt:
```diff
@@ -51,6 +51,20 @@ target_link_libraries(
 install(TARGETS mp-units EXPORT mp-unitsTargets)
 
+# C++ module
+if(CMAKE_EXPERIMENTAL_CXX_MODULE_CMAKE_API)
```
How is this being set? Shouldn't it be explicitly set somewhere to a proper UID?

JohelEGP replied:
You can.
But the good practice is to pass it to the CMake configuration command.
That's the place where it belongs,
as it is the user who knows their version of CMake,
and them who should opt-into this experimental feature.

The Libc++ module for `std` doesn't do it this way, however: <https://libcxx.llvm.org/Modules.html#using-in-external-projects>.
When I use that, I need to comment it out because it doesn't match CMake vNext.

mpusz replied:
Yes, I spoke with Bill Hoffman yesterday. He said the next release will remove the EXPERIMENTAL tag from C++ Modules.

mpusz commented:
> This is how I tested this (using [CPM](https://github.com/cpm-cmake/CPM.cmake) instead of Conan).

We need to be able to trigger this from Conan.

mpusz commented:
Hi @JohelEGP, it would be great to merge this PR but there are still some unresolved issues we discussed above.

JohelEGP commented:
All that's left is the integration with Conan.
I'll leave that to you.

I suppose you're going to add an option to Conan, which translates to a CMake option.
You should use that in place of `CMAKE_EXPERIMENTAL_CXX_MODULE_CMAKE_API`.

<https://github.com/fmtlib/fmt/releases/tag/10.1.1> is needed for the fixes for C++ modules.
Of <https://github.com/mpusz/mp-units/issues/7#issuecomment-1073138373>, <https://github.com/llvm/llvm-project/issues/54574> is still open, so warnings abound.

JohelEGP commented:
Unsurprisingly, GCC 13 and GCC 14 ICE.

mpusz commented on src/mp-units.cpp:
```diff
@@ -0,0 +1,69 @@
+#include <type_traits>
+#include <utility>
+
+// core-io
+#include <sstream>
+
+// core-fmt
+#include <mp-units/bits/fmt_hacks.h>
```
Shouldn't this be in separate submodules? Maybe there are some (e.g. embedded) projects that do not want or can't use this part?

mpusz commented on src/mp-units.cpp:
```diff
@@ -0,0 +1,69 @@
+
+export module mp_units;
+
+export
```
Should we export everything (even the `detail` namespace) or things like `detail_unit` that explicitly states:

https://github.com/mpusz/mp-units/blob/a356db749ef8651d6b8c48cf732673286a0faa39/src/core/include/mp-units/unit.h#L242-L246

JohelEGP replied:
Ideally, not.

JohelEGP commented on example/CMakeLists.txt:
```diff
@@ -30,7 +30,12 @@ target_include_directories(example_utils INTERFACE include)
     add_executable(${target} ${target}.cpp)
-    target_link_libraries(${target} PRIVATE ${ARGN})
+    if(TARGET mp-units::modules)
+        target_link_libraries(${target} PRIVATE mp-units::core-fmt mp-units::modules)
```
The implementation compatibility macros from the headers.

mpusz replied:
Shouldn't the macros be isolated similarly to the ones from core?

JohelEGP commented on docs/users_guide/examples/avg_speed.md:
````diff
@@ -18,28 +18,34 @@ First, we include all the necessary header files and import all the identifiers
 
 ```cpp title="avg_speed.cpp" linenums="1"
---8<-- "example/avg_speed.cpp:28:38"
+--8<-- "example/avg_speed.cpp:33:37"
+```
+```cpp title="avg_speed.cpp" linenums="6"
+--8<-- "example/avg_speed.cpp:28:29"
+```
+```cpp title="avg_speed.cpp" linenums="8"
+--8<-- "example/avg_speed.cpp:39:42"
````
I think users are unlikely to care for both.
They'll either use headers or modules.
If that's not the case,
we'd have to expose the macro to signal that (only) the modules are available.

mpusz replied:
So now the question is if we should show legacy or module-based versions of examples in our docs? 😉

mpusz commented:
The CI needs fixing as well. We should probably add modules tests to CI as well to check if there are no issues in the PR. Which compilers should the module-build support as of now? Only clang-16+?

When you are done with all the required changes and the CI will be fixed I will contribute to this PR with Conan-related changes to enable CI builds with modules.

mpusz commented on example/CMakeLists.txt:
```diff
@@ -24,13 +24,18 @@ cmake_minimum_required(VERSION 3.2)
     add_executable(${target} ${target}.cpp)
-    target_link_libraries(${target} PRIVATE ${ARGN})
+    if(TARGET mp-units::modules)
+        target_link_libraries(${target} PRIVATE mp-units::modules mp-units::core-fmt-modules mp-units::core-io-modules)
```
What about utilities?

JohelEGP replied:
I think the separation of core and utility is due to the standardization.
But maybe there's something lurking deeper.
Like `random.h` not being freestanding.

mpusz commented on .github/workflows/ci-conan.yml:
ci-test-package-cmake.yml should get similar changes as well.

mpusz commented on conanfile.py:
```diff
@@ -148,6 +148,7 @@ def generate(self):
         tc.variables["MP_UNITS_BUILD_LA"] = self._build_all and not self._skip_la
         tc.variables["MP_UNITS_USE_LIBFMT"] = self._use_libfmt
         if self._build_modules:
+            tc.variables["CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS"] = "/usr/bin/clang-scan-deps-16"
```
This looks fishy...

Is there any way to expose that through the CMake environment variables? Hardcoding this in a conanfile that might also be used on Windows seems wrong.

JohelEGP replied:
I agree.

First of all, this doesn't just work because the installed Clang package doesn't provide `clang-scan-deps` in `PATH`.
[`find_program`][] was probably used internally to do that,
and whatever its heuristics, didn't work for this.

[`find_program`]: https://cmake.org/cmake/help/latest/command/find_program.html

Normally, you'd just define this variable in a preset for CI or the CMake's command line.
You have an indirection through Conan, so I have no idea what's the equivalent way.

mpusz replied:
I think that a much better solution is to provide a custom CMake toolchain file with those variables set for CI and then call Conan in CI with the following:

```
conan -c tools.cmake.cmaketoolchain:user_toolchain=<path_to_toolchain> ...
```

This will prepend the Conan-generated toolchain file with the one provided by the user.

JohelEGP replied:
Good.
Now we just need to find out the path of the scanner
as installed by the Clang package.

mpusz replied:
```
$ which clang-scan-deps-16
/usr/bin/clang-scan-deps-16
```

Maybe we should generate this file on the fly in CI, as different versions of clang will require different variable values.

JohelEGP replied:
I already tried that (<https://github.com/mpusz/mp-units/actions/runs/6304075675/job/17114651127#step:14:692>):
> ` /bin/sh: 1: /usr/bin/clang-scan-deps-16: not found`

mpusz replied:
This should help: https://github.com/mpusz/mp-units/commit/993ea8e3ac8622f1f6647ae51a27cc256b60057e. It can be conditionally installed only when clang modules are needed.

mpusz replied:
You can easily add Conan `-c` options via "conan-config" GitHub Actions matrix parameter in _ci-conan.yml_.

mpusz commented on conanfile.py:
```diff
@@ -140,8 +144,12 @@ def layout(self):
         tc.variables["MP_UNITS_USE_LIBFMT"] = self._use_libfmt
+        if self._build_modules:
+            tc.variables["CMAKE_CXX_COMPILER_CLANG_SCAN_DEPS"] = "/usr/bin/clang-scan-deps-16"
+            tc.variables["CMAKE_EXPERIMENTAL_CXX_MODULE_CMAKE_API"] = "aa1f7df0-828a-4fcd-9afc-2dc80491aca7"
```
This also probably should land in a toolchain file.

JohelEGP replied:
<https://github.com/mpusz/mp-units/pull/350#discussion_r1336893398> is outdated but not resolved.

JohelEGP commented:
Looks like Clang 16 can't be used to test the modules.

mpusz commented:
> Looks like Clang 16 can't be used to test the modules.

I got the info from the  Conan team that Conan 2.0.12 should appear today and will have fixes to enable clang-17 configuration.

mpusz commented:
clang-17 support added with https://github.com/mpusz/mp-units/commit/92b7069c35bba592743c1132c4c2eccfa9c59c6c.

mpusz commented:
Please let me know when you will think it is ready to merge (i.e. in case you want to provide those toolchain files for CMake).

jcar87 commented:
Hi @mpusz - I've had a chance to test things out on this branch as I prepare to continue experiments with C++ module packaging.

A few comments:
* The `MP_UNITS_BUILD_MODULES` option is declared in the top-level `CMakeLists.txt` but is never evaluated - module functionality seems to be conditional on either of the following, which have some issues. 
    * `CMAKE_EXPERIMENTAL_CXX_MODULE_CMAKE_API` being defined. Note that this is no longer needed in CMake 3.28 - module scanning functionality is enabled automatically when C++20 mode is active (as far as I'm aware). 
    * or `if(TARGET mp-units::modules)` - this never entered the "true" branch for me when building this project with CMake 3.28 nightlies. Unsure why - this is ver dependent on CMake configuration order, but I can see the `src` directory is included before the others.
* My advice for the above would be to align everything around the `MP_UNITS_BUILD_MODULES` condition.

* Defining `cmake_minimum_required()` to an old version of CMake (e.g 3.4 or 3.5) - this was disabling the dependency scanning to that part of the tree - and I was getting a lot of `import` errors because CMake never ran clang-scan-deps for that file. Removing these lines did the trick. I'd say they're only needed at the top-level project, and any subdirectory that you expect to be "standalone" (that is, the root of a tree). Otherwise they're redundant at best, and dangerous at worse (since the call to `cmake_minimum_required()` changes the policy settings).  

* having (lots) of compiler errors with Clang 17 and GNU's `libstdc++` (clang on Linux uses the GNU standard library by default) - these could be due to issues specific to GNU's library, but worth cehcking

* using Clang 17 and `libc++` works without compiler errors, but I get linker errors instead - I suspect that this _may_ have to do things from #includes (standard library, GSL and fmt) being attached to the module. For everything that you do *not* want to be attached to the modules, you'll have to do an `#include` in the module preamble (after `module;` and before `export module xxxx;`. You can see an example here: https://github.com/fmtlib/fmt/blob/master/src/fmt.cc#L3-L36 - this will also apply to other things that are `#include`d , such as fmt (and maybe gsl?)

jcar87 commented:
As for the `cmake_minimum_required()` - indeed this will be a problem with CMake 3.28 if the different subdirectories have a call to that function for a version earlier than 3.28, as described here: 
* https://cmake.org/cmake/help/git-master/manual/cmake-cxxmodules.7.html
* https://cmake.org/cmake/help/git-master/policy/CMP0155.html#policy:CMP0155

There are a few alternatives:
* At the top level, evaluate `MP_UNITS_BUILD_MODULES` early on, and if true,  `set(CMAKE_CXX_SCAN_FOR_MODULES ON)` - this will enable dependency scanning for all targets, regardless

* Alternatively (and probably better): ensure the policy default is new, even for files that set `cmake_minimum_required()` for older versions: `set(CMAKE_POLICY_DEFAULT_CMP0155 NEW)`  - I think this ensures dependency scanning by default for all targets where C++20 or higher is used

* Removing calls to `cmake_minimum_required()` from files that don't need it (and inherit it from the parent) is probably the best approach. For the files that _do_ need it, may be safer to include 3.28, e.g. `cmake_minimum_required(3.5...3.28)`

mpusz commented:
I am still looking for the best modules structure for the project. There are a few alternatives:

# The simplest one

```mermaid
flowchart TD
    mp_units --- core
    mp_units --- systems
    mp_units --- ostream
    mp_units --- fmt
```

If someone does not want any I/O (i.e. embedded), the following is needed:

```cpp
import mp_units.core;
import mp_units.systems;
```

# Separate `no_io`

```mermaid
flowchart TD
    no_io --- core
    no_io --- systems
    mp_units --- ostream
    ostream --- no_io
    mp_units --- fmt
    fmt --- no_io
```

Now we just need:

```cpp
import mp_units.no_io;
```

but his somehow sucks 😉 

# Separate I/O options

```mermaid
flowchart TD
    mp_units --- core
    mp_units --- systems
    mp_units_ostream --- mp_units
    mp_units_fmt --- mp_units
```

In many cases, we use both I/O options (fmt and streams). In such a case, we need:

```cpp
import mp_units_ostream;
import mp_units_fmt;
```

# Unified I/O

```mermaid
flowchart TD
    mp_units --- core
    mp_units --- systems
    mp_units_ostream --- mp_units
    mp_units_fmt --- mp_units
    mp_units_io --- mp_units_ostream
    mp_units_io --- mp_units_fmt
```

Now we just need

```cpp
import mp_units_io;
```

mpusz commented:
I personally like the first one the most. It is the simplest one and allows selective inclusion of all the options.

mpusz commented:
We could consider breaking the project into even smaller pieces but I am unsure if it makes sense to provide separated modules for each system.

mpusz commented on .github/workflows/ci-conan.yml:
```diff
@@ -123,7 +122,7 @@ jobs:
             }
           - {
-              name: "Ubuntu Clang-17 + libc++ (modules)",
+              name: "Clang-17 (modules)",
```
You should also remove `std` line as it is not a part of the matrix.

JohelEGP commented:
I'm going to limit the scope to C++ and CMake sources.

mpusz commented:
@jcar87, could you please help @JohelEGP with his issues with Conan support for modules? We would like to merge this to the mainline ASAP.

JohelEGP commented:
I'd prefer to limit this to PR CMake.
It was working before.
But then I had to touch the Conan files, which I don't really understand.

Since the CMake buildsystems already work standalone,
the Conan support should be purely additive.

JohelEGP commented:
With regards to the first option of <https://github.com/mpusz/mp-units/pull/350#issuecomment-1763842790>,
the module `mp_units.core` would only include the headers `compare.h` and `random.h` from the `utility` sources.
The `utility` sources `chrono.h` and `math.h` require SI, so they would be part of `mp_units.systems`.

JohelEGP commented:
I suppose all that's missing now is documentation on how to use it.

You need to follow <https://cmake.org/cmake/help/latest/manual/cmake-cxxmodules.7.html>.
For the CMake buildsystems of the examples and tests in this repository,
which don't require a minimum CMake version of 3.28,
that means configuring with `-DCMAKE_CXX_SCAN_FOR_MODULES="TRUE"`.

Right now, if the CMake version is at least 3.28, the modules targets are unconditionally added.

mpusz commented:
Please see:

- c9d1f83ca493931909392cf28334e11dfd251ff9
- 7f0e7f8ca86c5860d6c3e5c62eaddcd43092a78a

mpusz commented on src/CMakeLists.txt:
```diff
@@ -51,6 +53,56 @@ target_link_libraries(
 install(TARGETS mp-units EXPORT mp-unitsTargets)
 
+# C++ modules
+if(CMAKE_VERSION VERSION_GREATER_EQUAL 3.28)
```
I would prefer to have a CMake option for C++ modules so we can test both modes (headers and modules) with the same CMake installed on a local or CI machine.

mpusz replied:
Of course, CMake could produce an error when MODULES opinion is selected, and the version is lower than required.

mpusz replied:
It already backfired, see: https://github.com/mpusz/mp-units/actions/runs/7263359786/job/19788472479?pr=350#step:7:267.

mpusz commented on src/core/CMakeLists.txt:
```diff
@@ -67,8 +66,7 @@ add_library(
     include/mp-units/system_reference.h
     include/mp-units/unit.h
 )
-target_compile_features(mp-units-core INTERFACE cxx_std_20)
```
Don't you think it is a good idea to set the minimum version requirements for the source code to be built in CMake? Yes, I know it is controversial as it overrides the user's setting and possibly results in a higher version than the one set by the user. But for that, we have `check_min_cppstd()` in the conanfile.py to prevent that.

If the user sets C++23 this will not be an issue as the user's setting will overwrite the value here as this one just states the minimum version required to compile those sources.

JohelEGP replied:
This was moved to `mp-units-core-macros`.

mpusz commented on src/core-macros/include/mp-units/bits/core-gmf.h:
```diff
@@ -0,0 +1,50 @@
+// core
+#include <gsl/gsl-lite.hpp>
+#include <mp-units/bits/external/hacks.h>
+#include <array>
+#include <compare>
+#include <concepts>
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+#include <initializer_list>
+#include <iterator>
+#include <limits>
+#include <numbers>
+#include <numeric>
+#include <optional>
+#include <ranges>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+// utility
+#include <chrono>
+#include <cmath>
+#include <random>
```
What is the status of `import std;` in clang?

JohelEGP replied:
You need to build it yourself: <https://libcxx.llvm.org/Modules.html#using-in-external-projects>.
It doesn't define `__cpp_lib_modules`, so you can't detect it.

mpusz commented on src/mp-units-systems.cpp:
```diff
@@ -0,0 +1,32 @@
+
+// utility
+#include <mp-units/chrono.h>
+#include <mp-units/compare.h>
+#include <mp-units/math.h>
+#include <mp-units/random.h>
```
`compare`, `random`, and `math_core`  should be in `mp-units-core`.

mpusz reviewed (changes requested):
Great work! Hopefully, we will have modules on the master branch soon 🎉

mpusz commented:
This commit should fix most of your CI issues: 8dae093517ab921cb53aaac698a4a2f2bb13e219.

mpusz commented:
If you do not have time for docs, you can leave updating them to me.

JohelEGP commented:
Rebased.

Added the option `MP_UNITS_BUILD_CXX_MODULES` to the CMake buildsystem and documentation.
Should we mention <https://cmake.org/cmake/help/latest/manual/cmake-cxxmodules.7.html>,
which documents the requirements for the support of C++ modules?
I wasn't sure where to put that, so I didn't include it.

Please, check whether commit 797752c724a9713250c5bf7569b98c3b74470d4c remains correct.

mpusz commented on docs/getting_started/installation_and_usage.md:
```diff
@@ -175,6 +175,15 @@ tools.build:compiler_executables={"c": "gcc-12", "cpp": "g++-12"}
+
+:   [:octicons-tag-24: 2.2.0][build_cxx_modules support] · :octicons-milestone-24: `ON`/`OFF` (Default: `OFF`)
+
+    Add C++ modules to the list of default targets.
```
I do not think it is correct. It says what it does, but now what is the outcome. From this description, it is not obvious that all examples and tests are built using modules.

mpusz commented on src/CMakeLists.txt:
```diff
@@ -28,6 +28,9 @@ set(projectPrefix MP_UNITS_)
 option(${projectPrefix}AS_SYSTEM_HEADERS "Exports library as system headers" OFF)
 message(STATUS "${projectPrefix}AS_SYSTEM_HEADERS: ${${projectPrefix}AS_SYSTEM_HEADERS}")
 
+option(${projectPrefix}BUILD_CXX_MODULES "Add C++ modules to the list of default targets" OFF)
```
The same comment from docs applies here. It is not obvious what it does. Maybe something like "Enables C++ modules-based build" would be better.

mpusz replied:
After a longer thought I think that this comment here is OK. The end user does not care how we build our tests and examples.
However, in the documentation, we can add a second sentence describing the outcome of this option for the local build.

mpusz commented on src/core/include/mp-units/bits/math_core.h:
Can we handle this file the same way we handle compare and random? If we move it to the core, it probably should not be in bits but in the parent directory with a file name that is more encouraging for direct use.

Some time ago we agreed that the purpose of keeping the `utility` directory is purely for standardization purposes. So that we know what can be potentially skipped in the first phase of the standardization and that the core headers do not depend on it. Otherwise, there is no sense in keeping the `utility` directory if there will be no `utility` module for it.

mpusz replied:
The above means that I am open to getting rid of the `utility` directory if we find it a better design and layout for the repo.

mpusz commented on src/utility/include/mp-units/math.h:
```diff
@@ -29,5 +29,5 @@
 // With C++20 modules:
-// - math_core will be a part of the mp_units.core module
-// - math_si and math_angular will be provided with the mp_units.systems module
+// - math_core is be a part of the mp_units.core module
```
```suggestion
// - math_core is a part of the mp_units.core module
```

mpusz commented:
> Should we mention https://cmake.org/cmake/help/latest/manual/cmake-cxxmodules.7.html,
which documents the requirements for the support of C++ modules?

Yes, I think it would be useful. We probably provide a bit more information about the modules hierarchy (the graph from earlier messaged in this PR) and how to build it. Maybe a new chapter somewhere on the "Installation and Usage" page (maybe before "Contributing (or just building all the tests and examples"). But you may leave it to me if you do not have time for it.

mpusz commented:
> Please, check whether commit https://github.com/mpusz/mp-units/commit/797752c724a9713250c5bf7569b98c3b74470d4c remains correct.

The best and fastest way to check docs is to serve a local instance of the docs with `mkdocs serve`. You can find more in https://mpusz.github.io/mp-units/2.2/getting_started/installation_and_usage/#building-documentation.

mpusz commented:
Also, could you please run `pre-commit run --all-files` to fix "Check CI" issues?

mpusz commented:
And of course, we need some unit tests for modules in the CI :wink:

JohelEGP commented:
But CI uses conan, which doesn't have an option corresponding to `MP_UNITS_BUILD_CXX_MODULES`.

mpusz commented:
As we are exposing regular targets as well with modules, I think that we might need two options in CMake and Conan. The first is to enable the building of modules. This could be set automatically as you tried, but we need better checks (compiler version, CMake version, build system type and version), which might be really hard to manage. The second one is to use the modules for not only a local build of tests and examples but also for the Conan customer projects.

JohelEGP commented:
> The second one is to use the modules for not only a local build of tests and examples but also for the Conan customer projects.

I don't know how you'd do it in Conan.
In CMake, you just use the new targets.

kobi-ca commented:
piggyback on this issue. I'm trying to run `CPMAddPackage` like this:
```cmake
CPMAddPackage("gh:mpusz/mp-units#f3b18e055fc96fb0cbe85fdbb96a39dac369051b")
```
and getting
```shell
.../build/_deps/mp-units-src/CMakeLists.txt'
  is meant to be used only as a CMake entry point and should not be included
  from other CMake files.  Include
  '.../build/_deps/mp-units-src/src/CMakeLists.txt'
  directly instead.
Call Stack (most recent call first):
  build/_deps/mp-units-src/CMakeLists.txt:40 (ensure_entry_point)
```

what am I doing wrong?

thanks!

mpusz commented:
Hi, the problem is that any solution that uses CMake's `add_subdirectory` for dependencies is not the proper way to do the dependencies management in C++. I spoke about it in one of my talks: https://youtu.be/mrSwJBJ-0z8?t=1931.

As stated in our [FAQ](
https://mpusz.github.io/mp-units/latest/getting_started/faq/#why-dont-we-have-cmake-options-to-disable-the-building-of-tests-and-examples) it is not only about tests and examples. There are many issues related to this approach.

I strongly recommend using a proper package manager like Conan. Alternatively, to fix your issue, you have to find a way to do `add_subdirectory` on the `./src` subdirectory rather than the main one in the repo.

mpusz commented:
This project's directory structure is described in our [docs](https://mpusz.github.io/mp-units/latest/getting_started/installation_and_usage/#repository-directory-tree-and-dependencies).

kobi-ca commented:
> This project's directory structure is described in our [docs](https://mpusz.github.io/mp-units/latest/getting_started/installation_and_usage/#repository-directory-tree-and-dependencies).

thanks. I can start working on bringing in `conan` for our project.

I tried this small test (just as a quick try before bedtime):
```cmake
# https://stackoverflow.com/questions/72362303/how-to-use-fetchcontent-populate-with-eigen
cmake_minimum_required (VERSION 3.12)

project (FetchContentExample)

# download CPM.cmake
file(
  DOWNLOAD
  https://github.com/cpm-cmake/CPM.cmake/releases/download/v0.38.3/CPM.cmake
  ${CMAKE_CURRENT_BINARY_DIR}/cmake/CPM.cmake
  EXPECTED_HASH SHA256=cc155ce02e7945e7b8967ddfaff0b050e958a723ef7aad3766d368940cb15494
)
include(${CMAKE_CURRENT_BINARY_DIR}/cmake/CPM.cmake)

CPMAddPackage("gh:gsl-lite/gsl-lite#bd9eb162d42d8ae6a7e86902ca7060247d71ac41")
CPMAddPackage("gh:fmtlib/fmt#7.1.3")


include (FetchContent)

FetchContent_Declare(
  mp-units
  GIT_REPOSITORY https://github.com/mpusz/mp-units.git
  GIT_TAG        v2.2.0
)

FetchContent_GetProperties(mp-units)
if(NOT mp-units_POPULATED)
  FetchContent_Populate(mp-units)

  add_subdirectory(${mp-units_SOURCE_DIR}/src ${mp-units_BINARY_DIR}/src EXCLUDE_FROM_ALL)
endif()

```

But I'm getting `export` issue. Not sure if I want to fight it :)

```shell
...
-- Configuring done (4.7s)
CMake Error: install(EXPORT "mp-unitsTargets" ...) includes target "mp-units-core" which requires target "fmt" that is not in any export set.
CMake Error in build/_deps/mp-units-src/src/CMakeLists.txt:
  export called with target "mp-units-core" which requires target "fmt" that
  is not in any export set.

```

mpusz commented:
This is one of the reasons that I recommend using true package managers. mp-units has its own dependencies as stated in our [docs](https://mpusz.github.io/mp-units/latest/getting_started/installation_and_usage/#repository-directory-tree-and-dependencies). Those may depend on other packages as well. If you do not want to resolve all of this manually I strongly recommend using a proper package manager like Conan.

A short intro to Conan can be found here https://mpusz.github.io/mp-units/latest/getting_started/installation_and_usage/#obtaining-dependencies. Please note that mp-units 2.2 was just released yesterday and ut may take a few days for it to be added to ConanCenter.

mpusz commented:
If you want to use the latest stable package, the easiest way would be to follow: https://mpusz.github.io/mp-units/latest/getting_started/installation_and_usage/#conan-cmake-live-at-head.
